fix(input_select): Decouple `input_select()` and `input_selectize()`

ErdaradunGaztea opened the pull request:
Fixes #1687

This PR follows the suggestions made in the issue and repurposes current `input_select()` implementation as `_input_select_impl()`. The new `input_select()` keeps the same signature but raises deprecation warnings upon using any of the three Selectize-related parameters.

Additionally, passing anything that's not a string (i.e. a HTML tag or `None`) as an option label raises another deprecation warning. The final goal would be to replace `_Choices = Mapping[str, TagChild]` with `_Choices = Mapping[str, str]`, but I guess that comes after a period of deprecation.

---

It's my first PR to this project, so if I missed some important part of the PR process, please let me know.

schloerke commented on shiny/ui/_input_select.py:
```diff
@@ -282,6 +340,17 @@ def _normalize_choices(x: SelectChoicesArg) -> _SelectChoices:
+def _contains_html(x: _SelectChoices) -> bool:
+    for v in x.values():
+        if isinstance(v, Mapping):
+            if _contains_html(cast(_OptGrpChoices, v)):
```
At this point, `v` is the `_Choices` values of `_OptGrpChoices` Mapping.

```suggestion
            if _contains_html(cast(_Choices), v)):
```

ErdaradunGaztea replied:
With your other suggestion, this one is not relevant anymore, but you are right, that should have been `_Choices`!

schloerke commented on shiny/ui/_input_select.py:
```diff
@@ -282,6 +340,17 @@ def _normalize_choices(x: SelectChoicesArg) -> _SelectChoices:
+def _contains_html(x: _SelectChoices) -> bool:
+    for v in x.values():
+        if isinstance(v, Mapping):
+            if _contains_html(cast(_OptGrpChoices, v)):
+                return True
```
While the typings wouldn't recurse forever, the recursive nature of the function gives me a little pause.

Could we instead perform it directly as we know the recursive depth should only ever be 1 and we'd only check for "not strings"?

```suggestion
            # Check the `_Choices` values of `_OptGrpChoices`
            for vv in v.values():
                if not is instance(vv, str):
                    return True
```

schloerke commented on shiny/ui/_input_select.py:
```diff
@@ -192,6 +195,55 @@ def input_select(
     * :func:`~shiny.ui.input_radio_buttons`
     * :func:`~shiny.ui.input_checkbox_group`
     """
+    if selectize:
+        warn_deprecated(
+            "`selectize` parameter of `input_select()` is deprecated. "
+            "Use `input_selectize()` instead of passing `selectize=True`."
+        )
+
+    if remove_button is not None:
+        warn_deprecated(
+            "`remove_button` parameter of `input_select()` is deprecated. "
+            "Use `input_selectize()` instead."
+        )
+
+    if options is not None:
+        warn_deprecated(
+            "`options` parameter of `input_select()` is deprecated. "
+            "Use `input_selectize()` instead."
+        )
```
For all three parameters, please change the parameter type to `MISSING_TYPE` and default the value to `DEPRECATED` ([link](https://github.com/posit-dev/py-shiny/blob/785197861d209e04a2498972143660aaa10f6a96/shiny/types.py#L48))

This will clearly signal that it should not be used. (Similar to how `lifecycle::deprecated()` is used in R packages)

To check if the value is present, use code similar to `if isinstance(selectize, MISSING_TYPE):` for the three cases.

Thank you

schloerke reviewed (changes requested):
> It's my first PR to this project

Thank you for your PRs!

Please be sure to add an entry in the CHANGELOG.md file linking to your PR. ([Related example](https://github.com/posit-dev/py-shiny/blob/785197861d209e04a2498972143660aaa10f6a96/CHANGELOG.md?plain=1#L841))

schloerke commented:
The app test failing in https://github.com/posit-dev/py-shiny/actions/runs/14320952734/job/40142214665?pr=1947#step:5:1519 is due to a change in this PR. 


```bash
pytest ./tests/playwright/ai_generated_apps/bookmark/input_select
```

Please update the test app or fix the helper method. Thank you!

ErdaradunGaztea commented:
> The app test failing in https://github.com/posit-dev/py-shiny/actions/runs/14320952734/job/40142214665?pr=1947#step:5:1519 is due to a change in this PR.
> 
> ```shell
> pytest ./tests/playwright/ai_generated_apps/bookmark/input_select
> ```
> 
> Please update the test app or fix the helper method. Thank you!

I made a mistake in the code, that's why it wasn't passing; fixed that. I've addressed all your other points, too. Please tell me if the updated approach with the `DEPRECATED` sentinel is how you do it, because there aren't any examples in the code to compare to (`navbar_options_resolve_deprecated()` is the only one).

schloerke commented on shiny/ui/_input_select.py:
```diff
@@ -2,7 +2,8 @@
 
-from ..types import Jsonifiable
+from .._deprecated import warn_deprecated
+from ..types import Jsonifiable, MISSING_TYPE, DEPRECATED
```
Lint
```suggestion
from ..types import DEPRECATED, MISSING_TYPE, Jsonifiable
```
